src/shared/components/exonsCharts/ExonsChartStore.ts

```diff
@@ -0,0 +1,242 @@
+     * @returns {number} return 0 when no width defined.
+     */
+    static getTotalWidth(exons: ExonRangeExt[]): number {
+        return exons.reduce((accumulator, e) => {
+            e.width = e.width ? e.width : 0;
+            return accumulator + e.width;
+        }, 0);
+    }
+
+    @computed
+    get computedTranscripts(): EnsemblTranscriptExt[] {
+        return (this.transcripts.result || [])
+            .map(t => {
+                t.exons = <ExonRangeExt[]> t.exons
+                    .sort(
+                        // sort exons by rank
+                        (exon1: ExonRangeExt, exon2: ExonRangeExt) => exon1.rank - exon2.rank
+                    )
+                    .map((e: ExonRangeExt) => {
+                        // add color
+                        e.fillColor = e.fillColor ? e.fillColor : t.fillColor;
+                        // calculate width of exon
+                        e.width = e.exonEnd - e.exonStart;
+                        return e;
+                    });
+                t.totalWidth = ExonsChartStore.getTotalWidth(<ExonRangeExt[]> t.exons);
+                t.deltaX = 0; // initialise delta for the x axis
+                return t;
+            })
+            .filter(t => t.isReferenceGene) // get only reference gene transcripts
+            .map(t => {
+                // calculate five primes total length
+                t.fivePrimeLength = !t.utrs ? 0 : t.utrs.reduce((accumulator, utr) => {
+                    // only calculate five prime
+                    let fivePrimeWidth = utr.type === UtrType.FivePrime ? utr.end - utr.start : 0;
+                    return accumulator + fivePrimeWidth;
+                }, 0);
+                // Add pfam domains data
+                t.pfamDomains = this.getPfamDomainDetails(t.pfamDomains);
+                return t;
+            })
+            .map(t => {
+                // Associate fusions with their reference gene transcipts
+                t.fusions = this.getFusionDetails(t);
+                return t;
+            });
+    }
+
+    getPfamDomainDetails(pfamDomainRangeExts: PfamDomainRangeExt[]): PfamDomainRangeExt[] {
+        return (pfamDomainRangeExts || [])
+        // do not include empty objects
+        .filter((pfam: PfamDomainRangeExt) => Object.keys(pfam).length > 0)
+        // add fill color and calculate widths
+        .map((pfam: PfamDomainRangeExt) => {
+            // retrieve label and desc from pfamDomains
+            const pfamDomain = (this.pfamDomains.result || []).find(
+                domain => domain.pfamAccession === pfam.pfamDomainId);
+            // add name and descriptions
+            pfam.name = pfamDomain ? pfamDomain.name : '';
+            pfam.description = pfamDomain ? pfamDomain.description : '';
+            // add color
+            pfam.fillColor = 'orange';
+            // calculate width of pfam
+            pfam.width = pfam.pfamDomainEnd - pfam.pfamDomainStart;
```

Comment:
again, you are modifying these objects in-place, which is not a good way of doing things. probably it's not what you're intending either. please see my other comments for instructions on how to improve this